src/Plugin/CheckImports.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | CheckImport plugin:
+--   This plugin checks if imported identifiers as unqualified
+--   exist and lists them.
+module Plugin.CheckImports
+  ( -- NOTE: The name "plugin" should be used to be called via GHC plugin mechanism.
+    plugin,
+  )
+where
+
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Char (isAlpha)
+import Data.Foldable (for_)
+import Data.IORef (readIORef)
+import Data.List (foldl', sort)
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Set (Set)
+import qualified Data.Set as S
+import GHC.Driver.Session (DynFlags, getDynFlags)
+import GHC.Plugins
+  ( CommandLineOption,
+    ModSummary,
+    Name,
+    Plugin (..),
+    defaultPlugin,
+    localiseName,
+    showSDoc,
+  )
+import GHC.Tc.Types (TcGblEnv (..), TcM)
+import GHC.Types.Name.Reader
+  ( GlobalRdrElt (..),
+    GreName (..),
+    ImpDeclSpec (..),
+    ImportSpec (..),
+  )
+import GHC.Unit.Module (ModuleName)
+import GHC.Utils.Outputable (Outputable (ppr))
+import Prelude hiding ((<>))
+
+plugin :: Plugin
+plugin =
+  defaultPlugin
+    { typeCheckResultAction = typecheckPlugin
+    }
+
+printPpr :: (Outputable a, MonadIO m) => DynFlags -> a -> m ()
+printPpr dflags = liftIO . putStrLn . showSDoc dflags . ppr
+
+formatName :: DynFlags -> Name -> String
+formatName dflags name =
+  let str = showSDoc dflags . ppr . localiseName $ name
+   in case str of
+        (x : _) ->
+          if isAlpha x
+            then str
+            else "(" ++ str ++ ")"
+        _ -> str
+
+formatImportedNames :: [String] -> String
+formatImportedNames names =
+  case fmap (++ ",\n") $ sort names of
+    l0 : ls ->
+      let l0' = "  ( " ++ l0
+          ls' = fmap ("    " ++) ls
+          footer = "  )"
+       in concat ([l0'] ++ ls' ++ [footer])
+    _ -> "  ()"
+
+mkModuleNameMap :: GlobalRdrElt -> [(ModuleName, Name)]
+mkModuleNameMap gre = do
+  spec <- gre_imp gre
+  case gre_name gre of
+    NormalGreName name -> do
```

Comment:
This must be a pattern rather than a genuine constructor, right? But I can't find it on Hoogle.

What is this pattern checking? What does it mean when it fails?

Reply:
The relevant info is here:
https://hackage.haskell.org/package/ghc-9.2.3/docs/GHC-Types-Name-Reader.html#t:GreName
The other case is FieldGreName which corresponds to record field label. Right now, this code does not process the record field label, but this should be marked as TODO. I will add the note. Thanks!